src/modules/constructRelationalGraph.js

```diff
@@ -0,0 +1,265 @@
+
+export type SelectionCondition = {[key: string]: any};
+
+export type Graph = {
+  [key: string]: {
+    edges: {[key: string]: Array<JoinCondition>},
+    selections: Array<SelectionCondition>,
+  },
+};
+
+export const AvailableTables = ['Doctor', 'Patient', 'Department'];
+
+export const ExpressionType = {
+  SELECTION: 'selection',
+  JOIN: 'join',
+  RELATION: 'relation',
+  AND: 'and',
+  CMP: 'cmp',
+};
+
+// Create a node of the table name with relevant fields
+const addNode = (
+  graph: Graph,
+  tableName: string,
+  selection?: ?SelectionCondition
+) => {
+  if (!(tableName in graph)) {
+    graph[tableName] = {
+      selections: [],
+      edges: {},
+    };
+  }
+  if (selection) {
+    graph[tableName]['selections'].push(selection);
+  }
+  return tableName;
+};
+
+// Add an edge for two tables if there's a valid join relation between them
+const addEdge = (
+  graph: Graph,
+  leftTable: string,
+  rightTable: string,
+  joinCondition: JoinCondition,
+  leftSelection?: ?SelectionCondition,
+  rightSelection?: ?SelectionCondition
+) => {
+  addNode(graph, leftTable, leftSelection);
+  addNode(graph, rightTable, rightSelection);
+  if (!(rightTable in graph[leftTable]['edges'])) {
+    graph[leftTable]['edges'][rightTable] = [];
+  }
+  if (!(leftTable in graph[rightTable]['edges'])) {
+    graph[rightTable]['edges'][leftTable] = [];
+  }
+  graph[leftTable]['edges'][rightTable].push(joinCondition);
+  graph[rightTable]['edges'][leftTable].push(joinCondition);
+};
+
+const conditionIsOnValidAttributes = (
+  attribute: string,
+  tables: Array<string>
+) => {
+  const columns: Array<string> = [];
+  for (const table of tables) {
+    switch (table) {
+      case 'Doctor':
+        columns.push(...Doctor.columns);
+        break;
+      case 'Patient':
+        columns.push(...Patient.columns);
+        break;
+      case 'Department':
+        columns.push(...Department.columns);
+        break;
+      default:
+        console.error('Invalid table name', table);
+        throw new Error('Invalid table name');
+    }
+  }
+  return columns.includes(attribute);
+};
+
+// function to check if a selection condition is a join condition
+// assuming that the selection condition is on a single table and join condition is between two tables
+const isValidJoinColumn = (attribute: string, tables: Array<string>) => {
+  if (attribute.includes('.')) {
+    const table = attribute.split('.')[0];
+    if (!tables.includes(table)) return false;
+  } else if (!conditionIsOnValidAttributes(attribute, tables)) {
+    return false;
+  }
+  return true;
+};
+
+export const getTableFromAttribute = (
+  attribute: string,
+  tables: Array<string>
+): string => {
+  if (attribute.includes('.')) {
+    return attribute.split('.')[0];
+  }
+  for (const table of tables) {
+    if (conditionIsOnValidAttributes(attribute, [table])) {
+      return table;
+    }
+  }
+  // this should never happen except for edge cases
+  return '';
```

Comment:
What edge cases are these?

Reply:
One edge case here can be `WHERE 1 = 1;` selection condition, where attribute 1 is not a column in one of the tables, in that case we just return an empty string to indicate that.

Reply:
Are there other edge cases here? If the only edge case here is when an "attribute" is not actually an attribute at all, it would helpful to be explicit here. `undefined` is also probably a better value to return than an empty string.

Also, looking at this now, I think "scalar" is probably a better name to refer to "attributes." By attribute it really seems like you mean columns. But in this case, the value passed in is not necessarily a column.